torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -622,12 +620,14 @@ def moe_forward(self, x, topk_ids, topk_weight):
         # the tokens in `gathered_tokens` are headed for. This part doesn't need
         # gradient.
         with torch.no_grad():
-            gatherd_idxs = np.zeros(shape=(gathered_tokens.shape[0],), dtype=np.int32)
-            s = 0
-            # TODO: remove `tolist()`
-            for i, k in enumerate(tokens_per_expert_group.tolist()):
-                gatherd_idxs[s : s + k] = i % self.experts_per_rank
-                s += k
+            gatherd_idxs = torch.arange(
+                tokens_per_expert_group.numel(),
+                device=tokens_per_expert_group.device,
+            )
+            gatherd_idxs = (
+                gatherd_idxs.repeat_interleave(tokens_per_expert_group)
+                % self.experts_per_rank
```

Comment:
nit: it seems `% self.experts_per_rank` can be moved to after the `torch.arange()` call, which may make the meaning clearer? i.e. we first get a bunch of local expert IDs, then we repeat_interleave. Not a big deal though.

Reply:
Done.